2단계 - 연관 관계 매핑 PR 요청드립니다.

earth-h opened the pull request:
안녕하세요, 민수님 :) 

JPA 2단계 PR 피드백 요청드립니다.

우선, 1단계 진행 시 피드백 주셨던 내용 참고하여서 테스트 메소드명 재수정 진행하였습니다 !

이번 단계에서 초점을 맞추어 보았던 부분은, 아래와 같습니다.

- 양방향 연관관계가 필요할지?
  - 질문과 답변 사이에 양방향 연관관계가 필요할지 고민을 하다가, 질문을 먼저 조회하고 해당 질문Id에 대한 답변 조회해도 괜찮지 않을까 하여 양방향 연관관계는 맺지 않았습니다.
- LAZY로 fetchType을 설정해야할지?
  - 질문이나 답변 조회 할 때에 누가 쓴 질문/답변인지를 체크할 것으로 보여 LAZY가 아닌 EAGER로 설정하였습니다.
- cascade 설정
  - 실무에서는, JPA를 간단하게만 사용해본 터라 토이프로젝트 시에 cascade.ALL로만.. 적용을 해봤는데요! ALL로 하면, REMOVE까지 함께 진행되는 것으로 보여서, 우선적으로 필요한 PERSIST에 대해서만 추가하였습니다.

감사합니다 :)

earth-h committed:
test(repository,domain): 테스트메소드명 피드백 반영

AnswerRepositoryTest
AnswerTest
DeleteHistoryRepositoryTest
QuestionRepositoryTest
QuestionTest
UserRepositoryTest
- 구체화된 클래스이름이나 변수명 사용하지 않도록 변경
  - README.md에 용어사전 생성 예정
- DynamicTest는 시나리오 기반이므로 테스트명 변경

earth-h committed:
docs(guide): README.md 업데이트

테스트코드에서 사용되는 용어 정리

earth-h committed:
docs(guide): README.md 업데이트

JPA : 2단계 - 연관 관계 매핑 요구사항 정리

earth-h committed:
docs(guide): README.md 업데이트

2단계 체크 필요 사항 추가

earth-h committed:
test(domain): 테스트 픽스쳐 생성

기존에 static final로 선언된 테스트 픽스쳐 대신, 정적 팩토리 메소드를 통해 생성할 수 있도록 추가

earth-h committed:
docs(guide): README.md 업데이트

JPA 관련 설정 필요 항목들 추가 정리

earth-h committed:
refactor(domain): Question에 대해 User 연관관계 매핑

Question
- 질문의 작성자에 대한 연관관계 매핑 추가
  - Question:User=N:1
  - cascade 속성은 PERSIST로 하여 Question이 영속 상태가 되면 User 역시 영속이 되도록 함
  - fetchType은 질문 조회 시, 질문 작성자 또한 함께 사용하는 경우가 많을 것이라 판단하여 EAGER로 진행

earth-h committed:
fix(domain): 테스트 픽스쳐 public static으로 변경

TestAnswerFactory
- public static으로 변경하여 정적 팩토리 메소드로 사용하고자 함

earth-h committed:
refactor(domain): Answer에 대해 Question, User 연관관계 매핑

Answer
- 답변의 작성자, 답변이 속한 질문과 관련된 연관관계 매핑 추가
  - Answer:Question=N:1
  - Answer:User=N:1
  - cascade 속성을 PERSIST로 하여 Answer가 영속이 되면 연관관계에 있는 entity 역시 영속이 되도록 함
  - fetchType은 답변 조회 시, 작성자나 해당 답변이 속한 질문 역시 함께 조회하는 경우가 많을 것 같아 EAGER로 진행

earth-h committed:
refactor(domain): DeleteHistory에 대해 User 연관관계 매핑

DeleteHistory
- 삭제이력과 유저간 연관관계 매핑
  - DeleteHistory:User=N:1
  - cascade 속성을 PERSIST로 하여 DeleteHistory가 영속이 되면 연관관계에 있는 entity 역시 영속이 되도록 함
  - 삭제이력은 주로 리스트로 여러개 뽑을 것으로 보이고, 특정 유저의 삭제이력으로 볼 때에 한 해 연관관계 대상 역시 조회되도록 LAZY 설정

earth-h committed:
test(domain): 생성한 텍스트 픽스쳐 사용하도록 테스트코드 리팩토링

earth-h committed:
refactor(domain): 사용하지 않는 setter 메소드 제거

earth-h committed:
refactor(domain): User에 대해 equals, hashCode 오버라이딩 추가

User
- Question과 Answer 작성자인지 여부를 판단할 때 equals() 사용
  - 영속성 컨텍스트를 비우고 다시 동일 객체를 조회하면, 내부 내용은 같아도 다른 객체를 반환해주기 때문에 equals 오버라이딩

earth-h committed:
docs(guide): README.md 업데이트

체크리스트 업데이트

mskangg commented on README.md:
```diff
@@ -1,5 +1,18 @@
+* DeleteHistory: 삭제이력
+* isDeleted: 삭제여부
+* userId: 유저Id
+
```
용어사전을 정리하셨군요 💯 
다만, 테스트만을 위한 용어사전이라기 보단 이 프로젝트에서 사용되는 모든 언어에 동일하게 적용된다고 생각하시면 됩니다!

mskangg commented on src/test/java/qna/domain/TestAnswerFactory.java:
```diff
@@ -0,0 +1,8 @@
+package qna.domain;
+
+public class TestAnswerFactory {
```
테스트픽스쳐 👍

mskangg commented on src/main/java/qna/domain/User.java:
```diff
@@ -76,41 +76,38 @@ public boolean isGuestUser() {
+            return false;
+        }
+        User user = (User) o;
+        return Objects.equals(getUserId(), user.getUserId()) && Objects.equals(getPassword(),
+                user.getPassword()) && Objects.equals(getName(), user.getName()) && Objects.equals(
+                getEmail(), user.getEmail());
```
Entity의 동등성 비교는 보통 식별이 가능한 id로 비교하는데, 전체 속성을 다 비교하신 이유가 있으신가요~?
궁금합니다!

earth-h replied:
User에는 기본키인 id와 각 User별로 unique한 값을 가지는 userId가 존재합니다.

이때, id는 생성전략이 IDENTITY여서 null일 수 있다고 판단하여서 id는 동등성 비교 대상에서 제외하였는데요 !
userId 외에 다른 필드는, 보다 정확하게 확인하려면 전체를 비교해야하지 않을까하여서 비교대상에 넣었습니다. 

남겨주신 댓글 보고 userId만으로도 동등성 비교가 충분할까하여서 equals() 오버라이딩을 id로 했을 때와 userId로 했을 때, 그리고 현재를 비교해보았습니다.
그 결과, userId만 있으면 동등성 비교에 문제가 없다고 판단하였습니다 !

한 번 더 체크할 수 있도록 리뷰해주셔서 감사합니다 :)

mskangg replied:
아 제가 말씀드린 부분은 userId가 아니였고 entity의 기본키인
```java
@Id @GeneratedValue(strategy = GenerationType.IDENTITY)
private Long id;
```
입니다!

jpa를 다루는 환경에서 entity는 영속성을 끌고 오냐 안오냐가 중요하기때문에 id가 null일 경우 영속성이 없음을 간주할 수 있고 동등비교도 id만으로 충분할 수 있기에 피드백 드렸던 내용이에요!
그렇기 때문에 id를 세팅하는 세터를 만들면 안되는 부분이 있죠
물론 id만 비교하는 경우 예외상황들이 충분히 발생하여 속성들을 추가하곤 하는데, 그렇기에 이유를 여쭤봤습니다 😁

earth-h replied:
https://velog.io/@park2348190/JPA-Entity의-equals와-hashCode

제가 참조하였던 글은 위 사이트인데요 :) 
기본키만으로 하려다가, 준영속 상태의 엔티티를 비교할 때 NPE가 발생할 여지가 있다고 판단하여서 진행했습니다..!

근데 지금 글을 다시 읽어보니, 오히려 오버라이딩하지 않고, JPA가 기본적으로 작동하는 equals를 사용하면 굳이 오버라이딩이 필요하지 않았을 수 있겠단 생각 드네요 😂

제가 테스트할 때에는 '오버라이딩을 한다'라는 조건 속에서 테스트하다보니 기본키로는 하면 안되겠구나 하는 결론을 냈던 것 같습니다..! 

피드백 감사합니다 !

mskangg replied:
제가 원하는 답변을 해주셨네요!
무슨 생각을 갖고 진행하셨는지, 그리고 준영속 상태의 문제를 생각하시고 하신건지가 궁금했어요~~ 💯

mskangg reviewed (changes requested):
1단계에 피드백 드린 부분 정말 완벽하게 반영하셨네요!! 💯 
테스트에 관한건 이번 미션 사항은 아니지만, 미션 요구사항들을 정말 잘해주셔서 피드백드렸더니,, 완벽히 수행하셨군요 👍 
전보다 훨씬 테스트코드가 가독성이 좋아졌습니다. 🚀

이번 미션은 연관관계매핑 미션인데요!
몇가지 피드백드렸으니 이번에도 지수님 생각이 담긴 코드반영 기대할게요~~

> 양방향 연관관계가 필요할지?
질문과 답변 사이에 양방향 연관관계가 필요할지 고민을 하다가, 질문을 먼저 조회하고 해당 질문Id에 대한 답변 조회해도 괜찮지 않을까 하여 양방향 연관관계는 맺지 않았습니다.

이번 단계는 연관 관계 매핑입니다!
질문을 먼저 조회하고 해당 질문 Id에 대한 답변을 조회하는 방식은 repository에게 조회를 요청하는 행위이죠.
하지만 연관관계를 매핑하면 QnaService에서 AnswerRepository의 의존성을 걷어낼 수 있는 장점이 있어보여요~
다만, 무한적인 객체그래프를 만드는건 지양해야하니 나중에 관계를 끊어내더라도 현재는 관계를 매핑해보시는건 어떠신가요?
관계를 맺을 때와 끊을 때의 차이점도 느끼실 수 있을 것 같아요 😃

> cascade 설정
실무에서는, JPA를 간단하게만 사용해본 터라 토이프로젝트 시에 cascade.ALL로만.. 적용을 해봤는데요! ALL로 하면, REMOVE까지 함께 진행되는 것으로 보여서, 우선적으로 필요한 PERSIST에 대해서만 추가하였습니다.

현재 REMOVE를 추가하게되면 답변을 삭제했을 경우 질문이 삭제될까요?? 🤔

earth-h committed:
docs(guide): README.md 업데이트

용어사전 설명 업데이트

earth-h committed:
refactor(domain): User의 equals, hashCode 오버라이딩 변경

User
- equals, hashCode 오버라이딩 시 기존에는 User Entity의 id를 제외한 모든 필드를 넣었으나, 굳이 필요하지 않은 필드까지 동등성 비교 시 넣지 않아도 됨 확인
  - UserTest 통해 확인
    - 영속성 컨텍스트 초기화 후 테스트 및 준영속 상태일 때 모두 테스트 진행한 결과, id로 동등성 비교하지 않고 userId로 동등성 비교하면 이슈없음 확인 완료

earth-h commented:
> 이번 단계는 연관 관계 매핑입니다!
질문을 먼저 조회하고 해당 질문 Id에 대한 답변을 조회하는 방식은 repository에게 조회를 요청하는 행위이죠.
하지만 연관관계를 매핑하면 QnaService에서 AnswerRepository의 의존성을 걷어낼 수 있는 장점이 있어보여요~
다만, 무한적인 객체그래프를 만드는건 지양해야하니 나중에 관계를 끊어내더라도 현재는 관계를 매핑해보시는건 어떠신가요?
관계를 맺을 때와 끊을 때의 차이점도 느끼실 수 있을 것 같아요 😃

말씀해주신것처럼, 이번 과제는 연관관계 매핑이니 여러가지 경우의 수를 다뤄보고자 Answer과 Question간에 양방향 매핑을 진행해보았습니다.
이후 진행할 질문 삭제 리팩토링 시, 양방향으로 진행하게 되면, Question 조회 시 Answer도 get 해 올 수 있는 장점이 확실히 있을 것 같습니다 :)
그 외에로는, 질문 삭제 관련해서 양방향의 이점이 있을까 싶어서(유저와 질문/답변사이) 해당 부분은 추가하지 않았습니다 !

양방향 매핑을 진행하면서, 무한루프 걸리지 않는지 테스트 진행해보면서 적용하였습니다.

==========================================================================================

> 현재 REMOVE를 추가하게되면 답변을 삭제했을 경우 질문이 삭제될까요?? 🤔
- 요 질문에 대해서 테스트해보고자 아래와 같이 테스트해보았습니다 :)

Answer Entity에 대해 Question 연관관계에 대해 cascade에 REMOVE 설정을 하니 Answer Entity 삭제 시 Question도 조회가 되지 않는 점 확인하였습니다 :) 

```java
@Entity
public class Answer extends BaseEntity {

    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    @ManyToOne(cascade = CascadeType.PERSIST)
    @JoinColumn(name = "writer_id", foreignKey = @ForeignKey(name = "fk_answer_writer"))
    private User writer;
    @ManyToOne(cascade = {CascadeType.REMOVE, CascadeType.PERSIST})
    @JoinColumn(name = "question_id", foreignKey = @ForeignKey(name="fk_answer_to_question"))
    private Question question;
    @Lob
    private String contents;
    @Column(nullable = false)
    private boolean deleted = false;
```

```java
    @Test
    void cascade_REMOVE_테스트() {
        //given
        User writer = TestUserFactory.create("javajigi");
        Question question = TestQuestionFactory.create(writer);
        Answer answer = TestAnswerFactory.create(writer, question);
        Answer saveAnswer = answerRepository.save(answer);
        Question saveQuestion = questionRepository.save(question);

        //when
        answerRepository.delete(saveAnswer);
        Optional<Question> findQuestion = questionRepository.findById(saveQuestion.getId());

        assertThat(findQuestion).isNotPresent();
    }
```

earth-h committed:
test(domain): AnswerRepositoryTest 테스트 케이스 추가

AnswerRepositoryTest
- Answer의 Question과의 연관관계에 cascade REMOVE를 넣었을 때 Answer 삭제 시 Question도 삭제되는지 테스트하고자 테스트코드를 작성

earth-h committed:
refactor(properties): SQL 보여주는 방식 logger로 변경

earth-h committed:
refactor(properties): SQL 바인딩값 보이도록 설정

earth-h committed:
refactor(domain): 양방향 매핑관계 설정

Answer
- 양방향 매핑관계에 따라 연관관계 편의 메소드(toQuestion) 내에 question entity의 answers 리스트 추가

Question
- 양방향 매핑관계 설정

AnswerTest
- 양방향 매핑관계에 따른 무한 루프 걸리지 않는지 테스트
  - toString()의 경우, Answer에서 Question 출력하게 하는 부분은 그대로 두고, Question에는 toString에 Answer 추가 안함

mskangg commented on src/main/java/qna/domain/Question.java:
```diff
@@ -18,9 +25,13 @@ public class Question extends BaseEntity {
+    private User writer;
     @Column(nullable = false)
     private boolean deleted = false;
+    @OneToMany(mappedBy = "question")
+    private List<Answer> answers = new ArrayList<>();
```
양방향 연관관계를 맺어주셨군요 👍

mskangg reviewed (approved):
> 기본키만으로 하려다가, 준영속 상태의 엔티티를 비교할 때 NPE가 발생할 여지가 있다고 판단하여서 진행했습니다..!

답글에도 남겨드렸지만, 무슨 생각을 갖고 진행하셨는지가 궁금했습니다. 😁
개인적으로 코드를 작성할 때 무작정 따라하는 것 보단 고찰 끝에 작성하는, 의도가 있는 코드를 좋아합니다 ㅎㅎ

이번 미션 잘 진행하셨어요 💯 
다음 단계도 화이팅입니다! 🚀
